File: problems/merge-sorted-lists/merge_sorted_lists.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

typedef struct ListNode {
  int val;
  uint32_t next;
} ListNode;

typedef struct Arena {
  ListNode *data;
  uint32_t size;
  uint32_t capacity;
} Arena;

Arena *arena_new(uint32_t capacity) {
  Arena *a = malloc(sizeof(Arena));
  a->data = malloc(sizeof(ListNode) * capacity);
  a->size = 0;
  a->capacity = capacity;
  return a;
}

uint32_t arena_alloc(Arena *a, ListNode n) {
  uint32_t idx = a->size++;
  a->data[idx] = n;
  return idx;
}

void arena_reset(Arena *a) {
  a->size = 0;
}

#define NULL_NODE 0xFFFFFFFF

uint32_t create_list(Arena *a, int len, int start, int step) {
  if (len <= 0)
    return NULL_NODE;
  uint32_t next = create_list(a, len - 1, start + step, step);
  ListNode node = {start, next};
  return arena_alloc(a, node);
}

uint32_t merge_two_lists(Arena *a, uint32_t l1, uint32_t l2) {
  if (l1 == NULL_NODE)
    return l2;
  if (l2 == NULL_NODE)
    return l1;

  if (a->data[l1].val < a->data[l2].val) {
    a->data[l1].next = merge_two_lists(a, a->data[l1].next, l2);
    return l1;
  } else {
    a->data[l2].next = merge_two_lists(a, l1, a->data[l2].next);
    return l2;
  }
}

int main() {
  int checksum = 0;

  Arena *arena = arena_new(1000);

  for (int i = 0; i < 5000; i++) {
    arena_reset(arena);
    uint32_t l1 = create_list(arena, 100, 0, 2);
    uint32_t l2 = create_list(arena, 100, 1, 2);
    uint32_t merged = merge_two_lists(arena, l1, l2);

    uint32_t curr = merged;
    while (curr != NULL_NODE) {
      checksum += arena->data[curr].val;
      curr = arena->data[curr].next;
    }
  }

  printf("Checksum: %d\n", checksum);
  return 0;
}

File: problems/merge-sorted-lists/merge_sorted_lists.rs

struct Arena {
    nodes: Vec<ListNode>,
}

impl Arena {
    fn new(capacity: usize) -> Self {
        Arena { nodes: Vec::with_capacity(capacity) }
    }
    fn alloc(&mut self, node: ListNode) -> u32 {
        let idx = self.nodes.len() as u32;
        self.nodes.push(node);
        idx
    }
    fn reset(&mut self) {
        self.nodes.clear();
    }
}

struct ListNode {
    val: i32,
    next: u32,
}

const NULL: u32 = u32::MAX;

fn create_list(arena: &mut Arena, len: i32, start: i32, step: i32) -> u32 {
    if len <= 0 {
        return NULL;
    }
    let next = create_list(arena, len - 1, start + step, step);
    arena.alloc(ListNode { val: start, next })
}

fn merge_two_lists(arena: &mut Arena, l1: u32, l2: u32) -> u32 {
    if l1 == NULL {
        return l2;
    }
    if l2 == NULL {
        return l1;
    }
    
    let v1 = arena.nodes[l1 as usize].val;
    let v2 = arena.nodes[l2 as usize].val;
    
    if v1 < v2 {
        let next = arena.nodes[l1 as usize].next;
        let merged_next = merge_two_lists(arena, next, l2);
        arena.nodes[l1 as usize].next = merged_next;
        l1
    } else {
        let next = arena.nodes[l2 as usize].next;
        let merged_next = merge_two_lists(arena, l1, next);
        arena.nodes[l2 as usize].next = merged_next;
        l2
    }
}

fn main() {
    let mut checksum = 0;
    let mut arena = Arena::new(500);
    for _ in 0..5000 {
        let l1 = create_list(&mut arena, 100, 0, 2);
        let l2 = create_list(&mut arena, 100, 1, 2);
        let merged = merge_two_lists(&mut arena, l1, l2);
        
        let mut curr = merged;
        while curr != NULL {
            checksum += arena.nodes[curr as usize].val;
            curr = arena.nodes[curr as usize].next;
        }
        arena.reset();
    }
    println!("Checksum: {}", checksum);
}
